### PR TITLE
Improved Support for Video Replay Support (including POST request canonicalization)

### LINKED ISSUE #80: Fuzzy Matching Improvements / POST requests
warc2zim now has a set of fuzzy matching rules (https://github.com/openzim/warc2zim/blob/master/src/warc2zim/main.py#L75) 
which are a subset of the larger ruleset in wabac.js (https://github.com/webrecorder/wabac.js/blob/main/src/fuzzymatcher.js#L8)

(pywb also has rules in python that are mostly aligned with the wabac.js rules https://github.com/webrecorder/pywb/blob/master/pywb/rules.yaml)

This many different rule sets is definitely a concern when it comes to maintenance, so perhaps should at least try to have  warc2zim use the wabac.js rules, since wabac.js is used for replay. These rules could be easily exposed as a json file that is loaded similar to the sw.js

Ideally, warc2zim would not need any rules and wabac.js could just read from zim using existing rules, but this is not possible for two issues:
- It is not possible to do a prefix query from the client. One of the ways that wabac.js handles replay is to do a a prefix search, usually by query string, eg. if an archived URL is `https://example.com/?A=B&_=1234` and there is a request for
`https://example.com/?A=B&_=1235`, it can do a prefix search for `https://example.com/?` and find the best match.
This approach allows for finding the best match URL from multiple possible captures.

Since prefix querying is not possible when loading from ZIM, the alternative is a custom canonicalization option, which wabac.js also supports: We create a  fake redirect', eg:

`https://example.fuzzy.replayweb.page/?A=B` which redirects to `https://example.com/?A=B&_=1234` in the ZIM
Then, when wabac.js encounters `https://example.com/?A=B&_=1235`, it also maps to `https://example.fuzzy.replayweb.page/?A=B`, and so is able to do the lookup.

This does work but is less flexible than the prefix search, as there is only possible match.

- A further complication of this is when POST request data is needed to lookup the URL, which is unfortunately now the case with youtube (as mentioned in webrecorder/browsertrix-crawler#4). Since WARCs contain the POST data,
wabac.js can take the POST data, especially if query or json and add it as part of the URL query.

For example, lets say a URL is the same but can only be distinguished by the POST data, which contains `{"videoid": "A"}`
A combined URL after reading the request and response can then be: `https://example.com/?_=1234&__post_json_data={"videoid": "A"}`, and the previous prefix search for `https://example.com/?` can find the best match.

For ZIMs, we'll need to do more work, though. The POST request must now also be parsed and a 'fake' redirect URL, probably something like `https://example.fuzzy.replaywebpage/?__post_json_data={"videoid": "A"}` generated.

This is doable, and can be added, but just wanted to raise awareness as this means creating (and continuing to maintain) a slightly different fuzzy matching scheme for ZIMs than exist for WARCs in wabac.js. The only possible alternatives, it seems, would be to allow for:
- providing a prefix API that could return a list of all ZIM records by prefix, eg: `https://example.com?`
- storing the WARC request data in case of POSTs

This issue is now coming up with youtube as youtube is making a POST request to the same URL, only difference is in the POST data (mentioned in webrecorder/browsertrix-crawler#4). The existing POST handling + prefix system means that replayweb.page is able to replay this new youtube playrer in WARCs + WACZ, but not ZIMs

Let me know if this makes sense, or can elaborate further..

### PR DESCRIPTION
**Pull request by ikreymer**
- Support video replay for current youtube and vimeo
- Use special iterate that matches request<->response pairs and includes POST request data from cdxj-indexer library.
- Update existing fuzzy matching rules for videos
- Utility: allow loading custom sw.js from a local path as well as URL
- Fixes #80

### CODE REVIEW

**Comment by kelson42** [comment]
@rgaudin Any feedback?

**Review by rgaudin — CHANGES_REQUESTED** [review]
Thanks for the PR @ikreymer !

Regarding warc2zim, it looks good to me except for the unpinned webac download in tests. We should definitely use fixed versions.

Unfortunately, the replayer is not working as expected. Here's what I get from [your kiwix-serve](https://kiwix-dev.webrecorder.net)

- Vimeo: video is *full sized* but the cover is blank (white). Playback works fine though.
- Procrastination: Gets proper cover/UI but playback fails to start (although the progress bar appears to be buffering some data)
Screen Shot 2021-11-01 at 23 49 34
- Youtube: same result but it's all in full-size (like Vimeo)
- Monabeach: all videos appear as grey boxes for a **very long time**. Then clicking play brings a black box for another long time.

When using the ZIM files, monabeach has a different behavior:
- Covers loads instantly on Safari
- Only the first three work. The other ones leads to black boxes.
- When clicking play, I can see a flashing kiwix-serve chrome (the toolbar) being displayed in the video box, before it goes dark.

I'd be happy to provide more inputs on something more specific. Maybe we should focus on the single-video tests first and once this works fine, expand to the more complex example

File: tests/test_warc_to_zim.py [hunk 1]
```diff
@@ -161,7 +197,7 @@ def test_warc_to_zim_specify_params_and_metadata(self, tmp_path):
                 "--zim-file",
                 zim_output,
                 "-r",
-                "https://cdn.jsdelivr.net/npm/@webrecorder/wabac@2.1.0-dev.3/dist/",
```
[the remark this answers is not in the document]
Comment (ikreymer) [reply tests/test_warc_to_zim.py:164]: Sure, updated to latest release, just checking for ability to download the file here.

**Comment by ikreymer** [comment]
Thanks for testing!
Unfortunately, I can't really repro the issues you are seeing. Will see if I can generate better samples, technically, these were created with archiveweb.page extension, not with zimit itself (since didn't have this ready yet).

Could you list the browsers and OS you are using to test? Perhaps there's a codec issue with Linux..

> Unfortunately, the replayer is not working as expected. Here's what I get from [your kiwix-serve](https://kiwix-dev.webrecorder.net)
> 
> * Vimeo: video is _full sized_ but the cover is blank (white). Playback works fine though.

I think that's what it is in this example?

> * Procrastination: Gets proper cover/UI but playback fails to start (although the progress bar appears to be buffering some data)
> 
Unfortunately, I can't repro this error, was able to get it to play in all three browsers but on OS X.

> 
> - Youtube: same result but it's all in full-size (like Vimeo)

Yes, this will be a full page as well as it using the embed directly. Maybe will remove this example.

 - Monabeach: all videos appear as grey boxes for a **very long time**. Then clicking play brings a black box for another long time.

Hmm, perhaps the initial load from this server is slow.

> When using the ZIM files, monabeach has a different behavior:
> 
> * Covers loads instantly on Safari
> * Only the first three work. The other ones leads to black boxes.
> * When clicking play, I can see a flashing kiwix-serve chrome (the toolbar) being displayed in the video box, before it goes dark.
> 

This is correct, only the first 3 videos are captured here (this was done manually again).

> I'd be happy to provide more inputs on something more specific. Maybe we should focus on the single-video tests first and once this works fine, expand to the more complex example

Sure, if you have additional examples, would be good to run some tests with the new zimit branch.
It sounds like main issue is with some of the single page youtube, maybe a capture via browsertrix-crawler would work better..

**Comment by ikreymer** [comment]
Here's an updated version of youtube: https://kiwix-dev.webrecorder.net/yt-2/
Can you see if this works, either hosted or locally?

**Comment by rgaudin** [comment]
> Here's an updated version of youtube: https://kiwix-dev.webrecorder.net/yt-2/ Can you see if this works, either hosted or locally?
> 
> Edit: Also added: https://kiwix-dev.webrecorder.net/yt-3/

Both work fine in Chrome and Safari on macOS, including when served via kiwix-serve with network disabled 👍 

I'm unable to get it working on Firefox though. Either I get the same error captured above or a black box…
There doesn't seem to be any error in the console (except a webfont and a favicon). Network tab shows a few (3) mp4 videoplayback chunks…


Ah and I spent a minute complaining about github not working again… forgot to re-enable my network 😅

**Comment by rgaudin** [comment]
Here's another round of feedback. Not different from earlier though. I am attaching a screencast so it's clearer for everybody.

https://user-images.githubusercontent.com/57929/141967980-fad8cbac-feb9-4d2d-9749-9cda1af61629.mp4

All the tests done on a **cleaned-up** Firefox DE 95.0b7 to make sure there is no cache or previous SW involved. Using DE specifically as FF is my regular browser. I'm testing on macOS Monterey.

Testing from https://kiwix-dev.webrecorder.net/

**yt-2**
- instant cover
- spinner for about 15s
- can see progress bar going a bit (matches with the two 206 requests)
- fails with error message and black video offers
- after a refresh, error stil happens but a lot quicker. About 1-2s

**yt-3**
- quick load of UI
- missing cover (black). This actually seem to vary based on window size. UI is probably trying to switch cover image based on size and we only have the scraper one.

**moab**
- relatively quick to load covers and UI
- video1 works OK
- clicking video2 leads to black box with no UI
- video3 works OK
- video4 to 14 black box

**vimeo**
- video frame is full width (not sure if that's expected ; I guess so)
- cover appears blurry while vimeo UI loads
- vimeo UI appears and cover disappears (white)
- video works OK

Now I am also attaching a screencast with Chrome where yt-2 works fine but the other issues are visible.


https://user-images.githubusercontent.com/57929/141969795-e891b637-7837-47e7-b484-c10ce1f2546c.mp4



@ikreymer could you please test with a **fresh** browser?

**Comment by ikreymer** [comment]
@rgaudin Thank you for the detailed videos, and sorry for late response.

The videos 4-14 on moab are not expected to work, they were never archived (this was a manual test). Sorry for the confusion!

I have the same results on chrome and I think everything is working as expected on chrome.

The only issues I see are on firefox:
- yt-2 should work. For me, it does, tried with fresh firefox
- the first 3 videos on moab should work.
- the loading of moab videos is very slow on firefox, perhaps the loading issue is related to that. I will see if I can track that down.

**Comment by rgaudin** [comment]
Thanks for the feedback ; looking forward to hear your findings. Ping me if there's anything I can do to help/test.

**Comment by ikreymer** [comment]
Unfortunately, it seems like there are some existing bugs with the Firefox service worker implementations, which results in requests just hanging and/or failing with `NS_BINDING_ABORTED` errors. I'm not quite sure when it happens, perhaps when there's a delay in responding, because it has to load from the upstream server. I've noticed that the requests that 'hang' are those that are duplicated (same URL loaded from multiple iframes), so maybe Firefox is getting confused by fetching same URL from upstream server.. not entirely sure. There are some related Firefox bugs such as:
https://bugzilla.mozilla.org/show_bug.cgi?id=1673534 which have suggested workaround to uninstall the service workers, which of course we can't do. The logging for firefox service worker is also a bit limited (https://bugzilla.mozilla.org/show_bug.cgi?id=1506892)

The problem seemed to be improved by enabling cacheing, so that multiple requests to the same `A/<url>` are cached and not fetched again. This seems to improve the replay in Firefox - the issues may still happen on first load, but after refresh, the pages do load and play for the examples below.

You can try this out in:
https://kiwix-dev.webrecorder.net/moab_2021-12/A/manonabeach.com/
and
https://kiwix-dev.webrecorder.net/yt-2_2021-12/

(be sure to clear cache and service workers or try new profile)

It still seems like there are issues with Firefox service workers, which is very hard to pinpoint, but this should be a bit better. (The cacheing is also similar to how loading from WACZ works, which does not have these issues).

**Comment by rgaudin** [comment]
Thank you for the reseach @ikreymer.

I can confirm that on moab, videos 1-3 works reliably. The cover/UI takes forever to loads though and , indeed, load instantly upon refresh.

On yt-2, there's no luck. It's still the same exact behavior.

You last comment regarding WACZ is unclear. Do you mean that WACZ works in the same way the ZIM now works (ie, better after a refresh) because of how content is served or do you mean that with WACZ, those videos works reliably on FF ?

**Comment by rgaudin** [comment]
@ikreymer ? Any answer for that last question ?

**Comment by ikreymer** [comment]
> On yt-2, there's no luck. It's still the same exact behavior.
> 

Hm, very odd, I'm unable to reproduce the issue, works on FF all versions for me.. Only thing I can think of is some sort of locale related difference..

> You last comment regarding WACZ is unclear. Do you mean that WACZ works in the same way the ZIM now works (ie, better after a refresh) because of how content is served or do you mean that with WACZ, those videos works reliably on FF ?

The second one. I meant that it seems to load reliably in FF from WACZ, and I suspect that is because the WACZ loading always caches to indexeddb, so avoids duplicate requests.

I think this might be as good as this can get for now.. Or do you want to try running additional tests with zimit of pages with youtube / vimeo?

**Comment by rgaudin** [comment]
Thanks ; our understanding also ; wanted to be sure about this WACZ difference but we'll merge and use this for now.

**Review by rgaudin — APPROVED** [review]

**Pull request merged**